File: aat-android/src/main/java/ch/bailu/aat/dispatcher/AndroidBroadcaster.java
package ch.bailu.aat.dispatcher;

import android.content.Context;
import android.content.Intent;

import java.util.HashMap;
import java.util.Map;

import ch.bailu.aat.util.AppIntent;
import ch.bailu.aat.util.OldAppBroadcaster;
import ch.bailu.aat_lib.dispatcher.BroadcastReceiver;
import ch.bailu.aat_lib.dispatcher.Broadcaster;
import ch.bailu.aat_lib.logger.AppLog;

public class AndroidBroadcaster implements Broadcaster {

    private final Context context;
    private final Map<BroadcastReceiver, android.content.BroadcastReceiver> observers = new HashMap<>(5);

    public AndroidBroadcaster (Context c) {
        this.context = c;
    }


    @Override
    public void broadcast(String action, Object... objects) {
        Intent intent=new Intent();
        intent.setAction(action);

        context.sendBroadcast(intent);
    }

    @Override
    public void register(BroadcastReceiver observer, String action) {
        if (!observers.containsKey(observer)) {
            observers.put(observer, new android.content.BroadcastReceiver() {
                @Override
                public void onReceive(Context context, Intent intent) {
                    observer.onReceive(AppIntent.toArgs(intent));
                }
            });
            OldAppBroadcaster.register(context, observers.get(observer), action);
        } else {
            AppLog.e(this, "Observer was allready registered.");
        }
    }

    @Override
    public void unregister(BroadcastReceiver observer) {
        android.content.BroadcastReceiver receiver = observers.remove(observer);
        if (receiver != null) {
            context.unregisterReceiver(receiver);

        } else {
            AppLog.e(this, "Observer was not registered.");
        }
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/dispatcher/AndroidBroadcaster.kt
package ch.bailu.aat.dispatcher

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import ch.bailu.aat.util.AppIntent
import ch.bailu.aat_lib.dispatcher.BroadcastReceiver
import ch.bailu.aat_lib.dispatcher.Broadcaster
import ch.bailu.aat_lib.logger.AppLog
import ch.bailu.foc.Foc

class AndroidBroadcaster(private val context: Context) : Broadcaster {
    private val observers: MutableMap<BroadcastReceiver, android.content.BroadcastReceiver> =
        HashMap(5)

    override fun broadcast(action: String, vararg args: String) {
        context.sendBroadcast(AppIntent.toIntent(action, *args))
    }

    override fun register(broadcastReceiver: BroadcastReceiver, action: String) {
        if (!observers.containsKey(broadcastReceiver)) {
            val receiver = object : android.content.BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {
                    broadcastReceiver.onReceive(*AppIntent.toArgs(intent))
                }
            }

            observers[broadcastReceiver] = receiver
            register(context, receiver, action)

        } else {
            AppLog.e(this, "Observer was already registered.")
        }
    }

    override fun unregister(broadcastReceiver: BroadcastReceiver) {
        val receiver = observers.remove(broadcastReceiver)
        if (receiver != null) {
            context.unregisterReceiver(receiver)
        } else {
            AppLog.e(this, "Observer was not registered.")
        }
    }


    companion object {
        fun broadcast(context: Context, action: String) {
            val intent = Intent()
            intent.action = action
            context.sendBroadcast(intent)
        }

        @JvmStatic
        fun register(context: Context, receiver: android.content.BroadcastReceiver, action: String) {
            val filter = IntentFilter(action)
            context.registerReceiver(receiver, filter)
        }

        fun broadcast(context: Context, action: String, file: Foc, url: String) {
            val intent = AppIntent.toIntent(action, file.path, url)
            context.sendBroadcast(intent)
        }
    }
}
